File: src/main/java/com/airsaid/localization/translate/impl/googleapi/GoogleApiTranslator.java
package com.airsaid.localization.translate.impl.googleapi;

import com.airsaid.localization.translate.AbstractTranslator;
import com.airsaid.localization.translate.TranslationException;
import com.airsaid.localization.translate.impl.google.AbsGoogleTranslator;
import com.airsaid.localization.translate.lang.Lang;
import com.airsaid.localization.translate.util.GsonUtil;
import com.google.auto.service.AutoService;
import com.intellij.openapi.util.Pair;
import com.intellij.util.io.RequestBuilder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * @author airsaid
 */
@AutoService(AbstractTranslator.class)
public class GoogleApiTranslator extends AbsGoogleTranslator {
  private static final String KEY = "GoogleApi";
  private static final String HOST_URL = "https://translation.googleapis.com";
  private static final String TRANSLATE_URL = HOST_URL.concat("/language/translate/v2");
  private static final String APPLY_APP_ID_URL = "https://cloud.google.com/translate";

  @Override
  public @NotNull String getKey() {
    return KEY;
  }

  @Override
  public @NotNull String getName() {
    return "Google (API)";
  }

  @Override
  public @NotNull String getRequestUrl(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text) {
    return TRANSLATE_URL;
  }

  @Override
  public String getAppKeyDisplay() {
    return "API Key";
  }

  @Nullable
  @Override
  public String getApplyAppIdUrl() {
    return APPLY_APP_ID_URL;
  }

  @Override
  public boolean isNeedAppId() {
    return false;
  }

  @Override
  public @NotNull List<Pair<String, String>> getRequestParams(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text) {
    List<Pair<String, String>> params = new ArrayList<>();
    params.add(Pair.create("q", text));
    params.add(Pair.create("target", toLang.getTranslationCode()));
    params.add(Pair.create("key", getAppKey()));
    params.add(Pair.create("format", "text"));
    return params;
  }

  @Override
  public void configureRequestBuilder(@NotNull RequestBuilder requestBuilder) {
    requestBuilder.tuner(connection -> connection.setRequestProperty("Referer", HOST_URL));
  }

  @Override
  public @NotNull String parsingResult(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text, @NotNull String resultText) {
    LOG.info("parsingResult: " + resultText);
    GoogleApiTranslationResult result = GsonUtil.getInstance().getGson().fromJson(resultText, GoogleApiTranslationResult.class);
    if (result.isSuccess()) {
      return result.getTranslationResult();
    } else {
      String message;
      if (result.error != null) {
        message = result.error.message.concat("(").concat(String.valueOf(result.error.code)).concat(")");
      } else {
        message = "Unknown error";
      }
      throw new TranslationException(fromLang, toLang, text, message);
    }
  }
}

File: src/main/kotlin/com/airsaid/localization/translate/impl/googleapi/GoogleApiTranslator.kt
package com.airsaid.localization.translate.impl.googleapi

import com.airsaid.localization.translate.AbstractTranslator
import com.airsaid.localization.translate.TranslationException
import com.airsaid.localization.translate.TranslatorCredentialDescriptor
import com.airsaid.localization.translate.impl.google.AbsGoogleTranslator
import com.airsaid.localization.translate.lang.Lang
import com.airsaid.localization.translate.util.GsonUtil
import com.google.auto.service.AutoService
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.Pair
import com.intellij.util.io.RequestBuilder

/**
 * @author airsaid
 */
@AutoService(AbstractTranslator::class)
class GoogleApiTranslator : AbsGoogleTranslator() {

    companion object {
        private val LOG = Logger.getInstance(GoogleApiTranslator::class.java)
        private const val KEY = "GoogleApi"
        private const val HOST_URL = "https://translation.googleapis.com"
        private const val TRANSLATE_URL = "$HOST_URL/language/translate/v2"
        private const val APPLY_APP_ID_URL = "https://cloud.google.com/translate"
    }

    override val key: String = KEY

    override val name: String = "Google"

    override fun getRequestUrl(fromLang: Lang, toLang: Lang, text: String): String = TRANSLATE_URL

    override val credentialDefinitions = listOf(
        TranslatorCredentialDescriptor(id = "appKey", label = "API Key", isSecret = true)
    )

    override val credentialHelpUrl: String? = APPLY_APP_ID_URL

    override fun getRequestParams(fromLang: Lang, toLang: Lang, text: String): List<Pair<String, String>> {
        return listOf(
            Pair.create("q", text),
            Pair.create("target", toLang.translationCode),
            Pair.create("key", credentialValue("appKey")),
            Pair.create("format", "text")
        )
    }

    override fun configureRequestBuilder(requestBuilder: RequestBuilder) {
        requestBuilder.tuner { connection -> connection.setRequestProperty("Referer", HOST_URL) }
    }

    override fun parsingResult(fromLang: Lang, toLang: Lang, text: String, resultText: String): String {
        LOG.info("parsingResult: $resultText")
        val result = GsonUtil.getInstance().gson.fromJson(resultText, GoogleApiTranslationResult::class.java)
        return if (result.isSuccess()) {
            result.translationResult
        } else {
            val message = if (result.error != null) {
                "${result.error!!.message}(${result.error!!.code})"
            } else {
                "Unknown error"
            }
            throw TranslationException(fromLang, toLang, text, message)
        }
    }
}
